analytical_engine/core/fragment/dynamic_fragment.h

```diff
@@ -2336,6 +2437,94 @@ class DynamicFragment {
     }
   }
 
+  // induce subgraph from induced_nodes
+  void induceFromVertices(std::shared_ptr<DynamicFragment>& origin,
+                          const std::unordered_set<oid_t>& induced_vertices,
+                          std::vector<edge_t>& edges) {
+    vertex_t vertex;
+    vid_t gid, dst_gid;
+    for (auto& oid : induced_vertices) {
+      if (origin->GetVertex(oid, vertex) && origin->IsInnerVertex(vertex)) {
+        // store the vertex data
+        vm_ptr_->GetGid(fid_, oid, gid);
+        vdata_[(gid & id_mask_)] = origin->GetData(vertex);
+        inner_vertex_alive_[(gid & id_mask_)] = true;
+
+        for (auto& e : origin->GetOutgoingAdjList(vertex)) {
+          auto dst_oid = origin->GetId(e.get_neighbor());
+          if (induced_vertices.find(dst_oid) != induced_vertices.end()) {
+            Oid2Gid(dst_oid, dst_gid);
+            edges.emplace_back(gid, dst_gid, e.get_data());
+          }
+        }
+        if (directed_) {
+          // filter the cross-fragment incoming edges
+          for (auto& e : origin->GetIncomingAdjList(vertex)) {
+            if (origin->IsOuterVertex(e.get_neighbor())) {
+              auto dst_oid = origin->GetId(e.get_neighbor());
+              if (induced_vertices.find(dst_oid) != induced_vertices.end()) {
+                Oid2Gid(dst_oid, dst_gid);
+                edges.emplace_back(dst_gid, gid, e.get_data());
+              }
+            }
+          }
+        }
+      }
+    }
+    // since we filtering alive vertex in vertex map construct, alive_ivnum
+    // equal to ivnum
+    alive_ivnum_ = ivnum_;
+  }
+
+  // induce edge_subgraph from induced_edges
+  void induceFromEdges(
+      std::shared_ptr<DynamicFragment>& origin,
+      const std::vector<std::pair<oid_t, oid_t>>& induced_edges,
+      std::vector<edge_t>& edges) {
+    vertex_t vertex;
+    vid_t gid, dst_gid;
+    edata_t edata;
+    for (auto& e : induced_edges) {
+      const auto& src_oid = e.first;
+      const auto& dst_oid = e.second;
+      if (origin->HasEdge(src_oid, dst_oid)) {
+        if (vm_ptr_->GetGid(fid_, src_oid, gid)) {
+          // src is inner
+          origin->GetVertex(src_oid, vertex);
```

Comment:
Using CHECK() to prevent the unexpected result of the function-call.

Reply:
update, thanks